### PR TITLE
promtail/server: Disable profiling by default

### LINKED ISSUE #5005: Add configuration option to disable cpu profiling endpoint in promtail
There is an undocumented(?) API endpoint in Promtail for CPU profiling: `/debug/fgprof`:
https://github.com/grafana/loki/blob/bd1fef08a8ea99e73a30c4739cfeb931353700e6/clients/pkg/promtail/server/server.go#L106

Considering https://github.com/felixge/fgprof/issues/12 one might want to disable this endpoint.
There should be a configuration option to disable it (or enable it if needed). There should also be documentation about this feature.

Yes, it can be done in an upstream reverse proxy, but having a configuration option would be better IMHO.

### PR DESCRIPTION
**Pull request by BenoitKnecht**
**What this PR does / why we need it**:

Add a `server.debug` option (that defaults to `false`) to enable the
`/debug/fgprof` endpoint in Promtail.

**Which issue(s) this PR fixes**:
Fixes #5005

**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: clients/pkg/promtail/server/server.go
Comment (cyriltovena) [inline clients/pkg/promtail/server/server.go:54]: I wonder if we should find a better name that reflect the behaviour.

There's other endpoint available for profiling under /debug/fgprof and this doesn't disable it. Should it ?

Comment (BenoitKnecht) [reply clients/pkg/promtail/server/server.go:54]: I didn't realize there was another endpoint under `/debug/`, but yes, I think it should be disabled as well. I chose this rather generic option name precisely because I figured there might be some other debugging facility (now or in the future) that should most likely not be running in production by default.

Comment (BenoitKnecht) [reply clients/pkg/promtail/server/server.go:54]: But I see the issue. The other profiling endpoint is `/debug/pprof`, which gets registered in a vendored dependency (https://github.com/grafana/loki/blob/main/vendor/github.com/weaveworks/common/server/server.go#L391-L397), along with `/metrics`.

So there isn't an easy way to disable `/debug/pprof` without disabling `/metrics` as well.

Comment (BenoitKnecht) [reply clients/pkg/promtail/server/server.go:56]: I've added a commit to deal with the `/debug/pprof` endpoint. Let me know what you think.

**Review by chaudum — COMMENTED** [review]
Instrumentation is enabled by default also in Loki. If we change the default (I don't recommend it for the same reason @cstyan mentioned) we'd want to change it there as well - which then also applies for other projects like Cortex or Tempo.

**Review by KMiller-Grafana — APPROVED** [review]
Docs part of this PR LGTM.
I agree with @chaudum about the parameter naming.

**Comment by cstyan** [comment]
> As for the default value, I'm surprised that there seems to be a majority of people arguing to keep it enabled by default. It seems very counter-intuitive to me that a software would run with debugging endpoints enabled by default. This is almost certainly not what most people want to do in production, and the principle of least surprise should apply here. For use cases where it's desirable to have them enabled, it can be done explicitly.

IMO, when running a distributed system like Loki there are issues that occur rarely, and having to explicitly enable instrumentation and then wait for that issue to pop up again in order to have the information you'll need in order to debug and diagnose is counter-intuitive.

**Comment by bboreham** [comment]
I think there is a bit of difference between server applications like Loki and Tempo, and promtail which is styled as an agent. You know you need to protect your servers from unwanted incoming calls; this may not be obvious for an agent. 

BTW I am a maintainer of `weaveworks/common` and can look at a PR to make profiling optional in the “server” package there.

**Comment by BenoitKnecht** [comment]
> I think there is a bit of difference between server applications like Loki and Tempo, and promtail which is styled as an agent. You know you need to protect your servers from unwanted incoming calls; this may not be obvious for an agent.

That's very true! In our case, Loki is behind an authenticating proxy, so the `/debug/*` endpoints are not accessible anyway, but `promtail` is deployed pretty much everywhere so it's quite impractical to restrict access to those endpoints systematically.

> BTW I am a maintainer of `weaveworks/common` and can look at a PR to make profiling optional in the “server” package there.

Are you saying you'd be open to PR submissions, or did you want to take a stab at it yourself?

**Comment by bboreham** [comment]
> Are you saying you'd be open to PR submissions, or did you want to take a stab at it yourself?

The former.

**Comment by cstyan** [comment]
Somehow during all of this I missed that this PR is for promtail. 

I would still say that we should keep the default as profiling on for other components (loki microservices) but feel less strongly about it for promtail itself.

**Comment by cstyan** [comment]
I think with an additional commit to fix the changelog conflict we could merge this?

**Comment by BenoitKnecht** [comment]
@cstyan I rebased on `main`, so the conflict should be fixed now.

**Comment by darxriggs** [comment]
please keep it open

**Comment by BenoitKnecht** [comment]
I've rebased again; could someone merge it, or is there anything else?

**Comment by BenoitKnecht** [comment]
@chaudum @cyriltovena Could you take a look? It would be great to get this merged soon.

**Comment by chaudum** [comment]
Seems that we agreed that we want to disable the profiling endpoints by default, so basically good to go.

However, regarding the configuration parameter naming, I would want to change it to `profiling_enabled`, with the `_enabled` suffix. I'm not sure if we are 100% consistent, but most of the config options that enable/disable stuff have the suffix as well.

Then I'm happy to merge the PR :)

**Comment by trevorwhitney** [comment]
@BenoitKnecht any updates? Would you be able to make those last few changes requested by @chaudum? Thanks!

**Comment by BenoitKnecht** [comment]
@trevorwhitney Sorry for the huge delay. I finally got around to make the requested changes. Let me know if there's anything else.

**Review by trevorwhitney — APPROVED** [review]
This looks good now, but I'd like a second opinion from @chaudum or @cstyan since they were more involved in the initial review. Thanks for the addition!

**Comment by BenoitKnecht** [comment]
I've rebased once again. Can we get this merged before it diverges again?

**Comment by BenoitKnecht** [comment]
@trevorwhitney There's been two releases of Loki since you merged this into `main`, but somehow this change isn't part of either of them. Is there something I need to do to get it pulled into the next release?

**Comment by bboreham** [comment]
Two patch releases.  It will be in the next minor release.

**Comment by trevorwhitney** [comment]
@BenoitKnecht FYI https://github.com/grafana/loki/issues/8797, looking like the end of next week, or early the week after for release.

**Comment by BenoitKnecht** [comment]
> @BenoitKnecht FYI #8797, looking like the end of next week, or early the week after for release.

Hey @trevorwhitney, that's great news, thanks for letting me know!

**Comment by otaku-bowei** [comment]
Execute ,I have a question
the commit is 404 not found,I want to know that —— if I want to disable pprof in loki configuration ,which version I should use

**Pull request merged**
